### DISCUSSION TITLE
How should MAX_VOLTAGE_TIME_SEC work?

### DISCUSSION THREAD

**Discussion by cflenker** [Louisvdw/dbus-serialbattery#870] [General]
I often observe the following behaviour:
When reaching **MAX_CELL_VOLTAGE** and the current is dropping towards zero the pack of high voltage cells come down and cross the pack of low voltage cells come up and end up below their voltage level.
At the short point of time when they are crossing the battery pack seems to be balanced an the timer of **MAX_VOLTAGE_TIME_SEC** starts.
So after **MAX_VOLTAGE_TIME_SEC** the control_voltage is lowered to FLOAT-Voltage.
In reality the battery is not completely balanced and **MAX_VOLTAGE** is stopped to early:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/125555670/e040d904-e5ab-40c2-9f0b-80b8b4178317)

I would prefer that voltagediff has to be lower than **CELL_VOLTAGE_DIFF_KEEP_MAX_VOLTAGE_UNTIL** for the whole time of **MAX_VOLTAGE_TIME_SEC**. If the voltagediff gets bigger than **CELL_VOLTAGE_DIFF_KEEP_MAX_VOLTAGE_UNTIL** the **MAX_VOLTAGE_TIME_SEC** should restart.

What Do you think about that?

I think the reason for the crossing is, that the cells with the bigger resistance come down as the current drops. When the current is close to zero you can see the real inbalance (but MAX_VOLTAGE_TIME_SEC already has started).

**Comment by mr-manuel** [comment]
A lot of batteries would never change to float in this way.

**Reply by cflenker** [reply]
Hello @mr-manuel,
I agree that there is a high risk that some batteries would never change to float this way. 
So maybe it is a possibility to offer a third threshold for celldiff to reset the MAX_VOLTAGE_TIME_SEC-counter if celldiff is above the threshold. Maybe you can have a look at https://github.com/Louisvdw/dbus-serialbattery/pull/895.
Thank you!
